pulp_docker/app/views.py

```diff
@@ -14,3 +57,235 @@ def get(self, request):
+        try:
+            ca = ContentArtifact.objects.get(
+                content__in=distribution.publication.repository_version.content,
+                relative_path=tag_name)
+            content = ca.content.cast()
+            if content.manifest:
+                headers = {'Content-Type': content.manifest.media_type}
+            else:
+                headers = {'Content-Type': content.manifest_list.media_type}
+
+        except MultipleObjectsReturned:
+            log.debug(
+                _('Multiple (pass-through) matches for {b}/{p}'),
+                {
+                    'b': distribution.base_path,
+                    'p': tag_name,
+                }
+            )
+
+        except ObjectDoesNotExist:
+            pass
+        else:
+            artifact = ca.artifact
+            if artifact:
+                return self._dispatch(artifact.file.name, headers)
+            else:
+                raise ArtifactNotFound(path)
+
+        raise PathNotResolved(path)
+
+
+class BlobManifestView(views.APIView, ServeContentMixin):
+    """
+    APIView for Docker Registry v2 blob and manifest APIs
+    """
+
+    authentication_classes = []
+    permission_classes = []
+    content_negotiation_class = IgnoreClientContentNegotiation
+
+    def get(self, request, path, digest):
+        """
+        Return a response to the "GET" action.
+        """
+        distribution = get_object_or_404(DockerDistribution, base_path=path)
+        try:
+            ca = ContentArtifact.objects.get(
+                content__in=distribution.publication.repository_version.content,
+                relative_path=digest)
+            headers = {'Content-Type': ca.content.cast().media_type}
+        except MultipleObjectsReturned:
+            log.debug(
+                _('Multiple (pass-through) matches for {b}/{p}'),
+                {
+                    'b': distribution.base_path,
+                    'p': hash,
+                }
+            )
+        except ObjectDoesNotExist:
+            pass
+        else:
+            artifact = ca.artifact
+            if artifact:
+                return self._dispatch(artifact.file.name, headers)
+            else:
+                raise ArtifactNotFound(path)
+
+        raise PathNotResolved(path)
+
+
+class TagsListView(views.APIView):
+    """
+    APIView for Docker Registry v2 tags/list API
+    """
+
+    authentication_classes = []
+    permission_classes = []
+    content_negotiation_class = IgnoreClientContentNegotiation
+
+    def get(self, request, path):
+        """
+        Return JSON of all tags in this repo
+        """
+
+        distribution = get_object_or_404(DockerDistribution, base_path=path)
+        tags = {'name': path, 'tags': []}
+        for c in distribution.publication.repository_version.content:
+            c = c.cast()
+            if isinstance(c, Tag):
+                tags['tags'].append(c.name)
+        return response.Response(tags)
```

Comment:
Should we file a story to paginate?

Reply:
we can file it as nice to have , but not every registry must have pagination.

Reply:
https://pulp.plan.io/issues/4168